### PR TITLE
Use invokedynamic instead of class generation for @ObjectHolder annotations

### PR DESCRIPTION
**Pull request by coderbot16**

### CODE REVIEW

File: src/main/java/com/patchworkmc/objectholder/ObjectHolder.java [hunk 1]
```diff
@@ -6,33 +6,26 @@
...
-		this.descriptor = entry.descriptor;
-		this.namespace = entry.namespace;
-		this.name = entry.name;
-	}
-
 	public String getField() {
 		return field;
 	}
 
-	public String getDescriptor() {
```
Comment (TheGlitch76) [inline src/main/java/com/patchworkmc/objectholder/ObjectHolder.java]: Is there really a need to only hide some of these?

File: src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java [hunk 2]
```diff
@@ -1,29 +1,53 @@
...
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
 
 import com.patchworkmc.Patchwork;
 import com.patchworkmc.annotation.StringAnnotationHandler;
+import com.patchworkmc.util.LambdaVisitors;
 
 public class ObjectHolderScanner extends ClassVisitor {
```
Comment (TheGlitch76) [inline src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java]: ObjectHolderRewriter, like the event one.

File: src/main/java/com/patchworkmc/access/AccessTransformation.java [hunk 3]
```diff
@@ -6,7 +6,6 @@
 	public static final AccessTransformation NONE = new AccessTransformation(0, 0);
 	public static final AccessTransformation DEFINALIZE = new AccessTransformation(Opcodes.ACC_FINAL, 0);
 	public static final AccessTransformation MAKE_PUBLIC = new AccessTransformation(Opcodes.ACC_PRIVATE | Opcodes.ACC_PROTECTED, Opcodes.ACC_PUBLIC);
-	public static final AccessTransformation DEFINALIZE_MAKE_PUBLIC = new AccessTransformation(Opcodes.ACC_FINAL | Opcodes.ACC_PRIVATE | Opcodes.ACC_PROTECTED, Opcodes.ACC_PUBLIC);
```
Comment (TheGlitch76) [inline src/main/java/com/patchworkmc/access/AccessTransformation.java:9]: why remove this? it's API, no need to prog:arr:d it

File: src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java [hunk 4]
```diff
@@ -1,29 +1,53 @@
...
+	// net.minecraft.util.Registry
+	private static final String REGISTRY = "net/minecraft/class_2378";
```
Comment (TheGlitch76) [inline src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java]: aren't these constants held in that registry util class?

File: src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java [hunk 5]
```diff
@@ -1,29 +1,53 @@
...
 	private boolean scanAllFields;
 	private String defaultModId;
+	private String className;
 
-	public ObjectHolderScanner(ClassVisitor parent, Consumer<ObjectHolder> consumer) {
+	public ObjectHolderScanner(ClassVisitor parent) {
 		super(Opcodes.ASM7, parent);
 
-		this.consumer = consumer;
+		holders = new ArrayList<>();
```
Comment (TheGlitch76) [inline src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java]: Just init this directly and make it `final`

**Review by TheGlitch76 — CHANGES_REQUESTED** [review]
Concept is good, some nitpicks as usual.

File: src/main/java/com/patchworkmc/access/AccessTransformation.java (hunk 3 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/access/AccessTransformation.java:9]: Eventually I'd like to completely remove/rework the builtin access transformer system, I'm just deleting unused stuff as I go. It seems like we'll only need to make classes public & definalize fields in the future, so this whole system will probably be dropped anyways in favor of something much simpler.

File: src/main/java/com/patchworkmc/objectholder/ObjectHolder.java (hunk 1 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/objectholder/ObjectHolder.java]: I think I forgot about this class, I was halfway through making it internal only.

File: src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java (hunk 2 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java]: :+1:

File: src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java (hunk 4 above)
Comment (coderbot16) [reply src/main/java/com/patchworkmc/objectholder/ObjectHolderScanner.java]: Oh yeah, I'll fix that.

File: src/main/java/com/patchworkmc/access/AccessTransformation.java (hunk 3 above)
Comment (TheGlitch76) [reply src/main/java/com/patchworkmc/access/AccessTransformation.java:9]: Alright, works for me.

**Review by TheGlitch76 — APPROVED** [review]
LGTM

**Comment by TheGlitch76** [comment]
next up, capabilities

**Pull request merged**